src/custom_component_renderer/book_directory_renderer.rs

```diff
@@ -0,0 +1,264 @@
+}
+
+/// Configuration from the book.toml file.
+pub struct RendererConfiguration {
+    pub i18n: I18nConfiguration,
+    pub destination: PathBuf,
+    pub current_language: Option<String>,
+    pub root: PathBuf,
+}
+
+impl RendererConfiguration {
+    pub fn new(
+        i18n: I18nConfiguration,
+        destination: PathBuf,
+        current_language: Option<String>,
+        root: PathBuf,
+    ) -> Self {
+        RendererConfiguration {
+            i18n,
+            destination,
+            current_language,
+            root,
+        }
+    }
+}
+
+pub struct RenderingContext {
+    pub path: PathBuf,
+    pub destination: PathBuf,
+    pub language: String,
+}
+
+impl RenderingContext {
+    fn new(path: PathBuf, destination: PathBuf, language: String) -> Self {
+        RenderingContext {
+            path,
+            destination,
+            language,
+        }
+    }
+
+    pub fn rendered_path(&self) -> String {
+        let mut relative_path = PathBuf::from(
+            self.path
+                .strip_prefix(&self.destination)
+                .expect("Invalid path"),
+        );
+        if let Ok(stripped) = relative_path.strip_prefix(&self.language) {
+            relative_path = stripped.to_owned();
+        }
+        String::from(
+            relative_path
+                .to_str()
+                .expect("Failed to convert path to rendered path"),
+        )
+    }
+}
+
+pub(crate) struct BookDirectoryRenderer {
+    config: RendererConfiguration,
+    components: Vec<Box<dyn Component>>,
+    languages_paths: BTreeMap<String, PathBuf>,
+}
+
+impl BookDirectoryRenderer {
+    pub(crate) fn new(config: RendererConfiguration) -> BookDirectoryRenderer {
+        let default_language = config.i18n.default_language.clone();
+        let languages_paths = config
+            .i18n
+            .languages
+            .keys()
+            .filter(|language| {
+                default_language.is_none() || *language != default_language.as_ref().unwrap()
+            })
+            .map(|language| {
+                (
+                    language.clone(),
+                    config.destination.join("html").join(language),
+                )
+            })
+            .collect::<BTreeMap<String, PathBuf>>();
+        BookDirectoryRenderer {
+            config,
+            languages_paths,
+            components: Vec::new(),
+        }
+    }
+
+    pub fn translate(&self) -> Result<()> {
+        let default_language = &self.config.i18n.default_language;
+        for (identifier, _) in &self.config.i18n.languages {
+            if let Some(default_language) = default_language {
+                if default_language == identifier {
+                    continue;
+                }
+            }
+
+            let destination = self.config.destination.as_path().display().to_string();
+            let book_folder = self.config.root.as_path().display().to_string();
+
+            Command::new("mdbook")
```

Comment:
This should not call out to a `mdbook` binary — what if people run `mdbook` as `cargo run` from a `mdbook` clone? If we need to invoke `mdbook`, we should instead invoke the functionality using the library functions in `mdbook`, similar to how `mdbook-i18n` does it.

However, I'm not sure we need to invoke `mdbook` in the first version: it would already be super helpful if we keep the renderer as a per-language renderer.

The simple for-loop outside of `mdbook build` is not so bad, and #13 is about making a wrapper tool which can call `mdbook build` for all languages. This tool could also handle `mdbook serve` then (probably by building all languages and starting a simple webserver).

Reply:
> This should not call out to a mdbook binary — what if people run mdbook as cargo run from a mdbook clone? If we need to invoke mdbook, we should instead invoke the functionality using the library functions in mdbook, similar to how mdbook-i18n does it.

It's really how `mdbook` itself does it, so I think it's fine. But I will anyway replace this child process with simple library calls, which will give more control, performance and stability. I will remove the need to spawn new processes by directly calling `book.clone()` -> `get_text(book)` -> `book.build()` but I first need to move the methods in `get_text` out into their own library

> However, I'm not sure we need to invoke mdbook in the first version: it would already be super helpful if we keep the renderer as a per-language renderer.

I can take it out of the first version. We'll see how it pans out in the end, my goal is to support both per language and global languages builds, so that it will be nicer for translators to debug directly from `mdbook serve`.

> The simple for-loop outside of mdbook build is not so bad, and https://github.com/google/mdbook-i18n-helpers/issues/13 is about making a wrapper tool which can call mdbook build for all languages. This tool could also handle mdbook serve then (probably by building all languages and starting a simple webserver).

My next commit will fix https://github.com/google/mdbook-i18n-helpers/issues/13 as well. The added benefits, beyond having one less step in github workflows, is that all the configurations will live in one place (`book.toml`), and `mdbook serve` will actually serve all the translations. Currently, if we want to add a new language, we need to update the workflow as well as `index.hbs`.  With this change, all would be handled for us and we would only need to care about the actual translation work.

## Next step

This week (probably Friday, unless I get some time before), I will replace the `Command::new` call with library calls. I will also add the `custom components` feature. Then we can see how we split the PR and decide  what features we don't want. My main concern is how complex the code will become, I think it will be ~800 LOC without tests and `Cargo.lock` (next commit will be around 80 LOC + moving ~100 LOC from `bin/get_text.rs` to `lib.rs`.)